pkg/liqo-controller-manager/namespaceMap-controller/namespaceMap_controller.go

```diff
@@ -0,0 +1,98 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespaceMap_controller
+
+import (
+	"context"
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+	const_ctrl "github.com/liqotech/liqo/pkg/liqo-controller-manager"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutils "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type NamespaceMapReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	Mapper        meta.RESTMapper
+	RemoteClients map[string]client.Client
+}
+
+const (
+	clusterIdForeign = "discovery.liqo.io/cluster-id"
+)
+
+func (r *NamespaceMapReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+
+	namespaceMap := &mapsv1alpha1.NamespaceMap{}
+	if err := r.Get(context.TODO(), req.NamespacedName, namespaceMap); err != nil {
+		klog.Errorf("%s --> Unable to get NamespaceMap '%s'", err, req.Name)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !ctrlutils.ContainsFinalizer(namespaceMap, const_ctrl.NamespaceMapControllerFinalizer) {
+		ctrlutils.AddFinalizer(namespaceMap, const_ctrl.NamespaceMapControllerFinalizer)
+		if err := r.Patch(context.TODO(), namespaceMap, client.Merge); err != nil {
+			klog.Errorf("%s --> Unable to add '%s' to the NamespaceMap '%s'", err, const_ctrl.NamespaceMapControllerFinalizer, namespaceMap.GetName())
+			return ctrl.Result{}, err
+		}
+		klog.Infof("Finalizer correctly added on NamespaceMap '%s'", namespaceMap.GetName())
+	}
+
+	if err := r.manageRemoteNamespaces(namespaceMap); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(namespaceMap.Status.CurrentMapping) == 0 {
```

Comment:
Why are you removing the finalizer if the status has no current mappings? Shouldn't you check the deletionTimestamp?

Reply:
I saw the update handler in the eventFilter, I got it.

Reply:
This finalizer is only used to ensure that the NamespaceMap is not deleted until there are remote Namespaces onto the corresponding remote clusters. The deletion logic of NamespaceMap is managed by Virtual-Node Controller, it will check deletion timestamp... I don't know if it is clear. There are 2 finalizers on NamespaceMap one setted by VirtualNode Controller and this one. If the NamespaceMap is ready for deletion and the virtualNode's finalizer is already been removed, we couldn't delete really the NamespaceMap until the are remote Namespaces, this is the scope of this new finalizer.

Reply:
Great, thank you for the review